hs_app_timeseries/models.py

```diff
@@ -44,16 +116,63 @@ class Variable(AbstractMetaDataElement):
 
-    class Meta:
-        # variable element is not repeatable
-        unique_together = ("content_type", "object_id")
+    @classmethod
+    def update(cls, element_id, **kwargs):
+        element = cls.objects.get(id=element_id)
+
+        # if the user has entered a new variable name
+        variable_name_str = 'variable_name'
+        if variable_name_str in kwargs:
+            if element.variable_name != kwargs[variable_name_str]:
+                # check if the user has entered a new name
+                if kwargs[variable_name_str] not in [item.name for item in
+                                                     element.metadata.cv_variable_names.all()]:
+                    # generate term for the new name
+                    kwargs[variable_name_str] = kwargs[variable_name_str].strip()
+                    term = _generate_term_from_name(kwargs[variable_name_str])
+                    variable_name = CVVariableName.objects.create(
+                        metadata=element.metadata, term=term, name=kwargs[variable_name_str])
+                    variable_name.is_dirty = True
+                    variable_name.save()
+
+        # if the user has entered a new variable type
+        variable_type_str = 'variable_type'
+        if variable_type_str in kwargs:
+            if element.variable_type != kwargs[variable_type_str]:
+                # check if the user has entered a new type
+                if kwargs[variable_type_str] not in [item.name for item in
+                                                     element.metadata.cv_variable_types.all()]:
+                    # generate term for the new name
+                    kwargs[variable_type_str] = kwargs[variable_type_str].strip()
+                    term = _generate_term_from_name(kwargs[variable_type_str])
+                    variable_type = CVVariableType.objects.create(
+                        metadata=element.metadata, term=term, name=kwargs[variable_type_str])
+                    variable_type.is_dirty = True
+                    variable_type.save()
+
+        # if the user has entered a new speciation
+        speciation_str = 'speciation'
+        if speciation_str in kwargs:
+            if element.speciation != kwargs[speciation_str]:
+                # check if the user has entered a new speciation
+                if kwargs[speciation_str] not in [item.name for item in
+                                                  element.metadata.cv_speciations.all()]:
+                    # generate term for the new name
+                    kwargs[speciation_str] = kwargs[speciation_str].strip()
+                    term = _generate_term_from_name(kwargs[speciation_str])
+                    speciation = CVSpeciation.objects.create(metadata=element.metadata, term=term,
+                                                             name=kwargs[speciation_str])
+                    speciation.is_dirty = True
+                    speciation.save()
```

Comment:
@pkdash Should setting ```is_dirty``` to True be done after ```save()``` call just in case there could potentially be an issue in ```save```?